### PR TITLE
Samgeo with preprocessing pipeline

### PR DESCRIPTION
**Pull request by joleitner**
To be able to preprocess the images all preprocessing steps had to get adjusted to handle Tif files. A final samgeo model was developed.

### CODE REVIEW

File: image-processing/core/preprocessing.py [hunk 1]
```diff
@@ -47,3 +47,64 @@ def remove_shadows(image: str | cv2.typing.MatLike, output: Optional[str] = None
...
+    res = cv2.cvtColor(hsv_image, cv2.COLOR_HSV2BGR)
+
+    if output is not None:
+        _save_image(res, output, tif_meta)
+
+    return res
+
+
+
+def _load_image(image: str | np.ndarray):
```
Comment (ProfBits) [inline image-processing/core/preprocessing.py]: Missing return type hint. Sould be tuple[np.ndarray, Optional[<metadata type>]] or so

File: image-processing/core/preprocessing.py [hunk 2]
```diff
@@ -47,3 +47,64 @@ def remove_shadows(image: str | cv2.typing.MatLike, output: Optional[str] = None
...
+                meta = src.meta.copy()
+                # Read with Rasterio and transpose to OpenCV format:
+                # from (bands, height, width) to (height, width, bands)
+                image = src.read().transpose(1, 2, 0)
+        else:
+            image = cv2.imread(image)
+
+    return image, meta
+
+def _save_image(res: np.ndarray, output: str, tif_meta: dict):
```
Comment (ProfBits) [inline image-processing/core/preprocessing.py]: Same here missing return type hint for None and tif_meta should be an Optional[dict]

File: image-processing/cut_houses/cut_houses.py [hunk 3]
```diff
@@ -224,9 +224,18 @@ def create_house_masks(image:str, output:str = None,)->np.ndarray:
 def cut_mask_from_image(image:str|np.ndarray, mask:str|np.ndarray, output:str = None, inverted:bool = False)-> np.ndarray:
 
     try:
+        # copy from core/preprocessing _load_image -> prevent import errors
```
Comment (ProfBits) [inline image-processing/cut_houses/cut_houses.py]: Shoulded than save and load moved to their own file and imported twice?

File: image-processing/cut_houses/cut_houses.py [hunk 4]
```diff
@@ -242,9 +251,18 @@ def cut_mask_from_image(image:str|np.ndarray, mask:str|np.ndarray, output:str =
         # cv2.imshow('Cutout Image Combined', cutout_image)
         # cv2.waitKey(0)
 
-        if(output != None):
+        if output is not None:
             try:
-                cv2.imwrite(output, cutout_image)
+                # copy from core/preprocessing _save_image -> prevent import errors
```
Comment (ProfBits) [inline image-processing/cut_houses/cut_houses.py]: Same here, why not move load and save to own file and import twice and adjust __init_.py

**Review by ProfBits — CHANGES_REQUESTED** [review]

File: image-processing/core/preprocessing.py (hunk 2 above)
Comment (joleitner) [reply image-processing/core/preprocessing.py]: erledigt

File: image-processing/core/preprocessing.py (hunk 1 above)
Comment (joleitner) [reply image-processing/core/preprocessing.py]: erledigt

File: image-processing/cut_houses/cut_houses.py (hunk 3 above)
Comment (joleitner) [reply image-processing/cut_houses/cut_houses.py]: hab ich jetzt so gemacht. Ich hoffe nur, dass niemand anderes jetzt import errors bekommt, weil ers von nem anderen pfad aus aufruft..

**Comment by joleitner** [comment]
@ProfBits Changes are made

**Review by ProfBits — APPROVED** [review]

**Pull request merged**